File: LeetCode/0208.go
package leetcode

type Trie struct {
	isword bool
	child  []*Trie
}

/** Initialize your data structure here. */
func Constructor() Trie {
	return Trie{false, make([]*Trie, 26)}
}

/** Inserts a word into the trie. */
func (this *Trie) Insert(word string) {
	cur := this
	for _, ch := range word {
		idx := ch - 'a'
		if cur.child[idx] == nil {
			cur.child[idx] = &Trie{false, make([]*Trie, 26)}
		}
		cur = cur.child[idx]
	}
	cur.isword = true
}

/** Returns if the word is in the trie. */
func (this *Trie) Search(word string) bool {
	cur := this
	for _, ch := range word {
		idx := ch - 'a'
		if cur.child[idx] == nil {
			return false
		}
		cur = cur.child[idx]
	}
	return cur.isword
}

/** Returns if there is any word in the trie that starts with the given prefix. */
func (this *Trie) StartsWith(prefix string) bool {
	cur := this
	for _, ch := range prefix {
		idx := ch - 'a'
		if cur.child[idx] == nil {
			return false
		}
		cur = cur.child[idx]
	}
	return true
}

/**
 * Your Trie object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(word);
 * param_2 := obj.Search(word);
 * param_3 := obj.StartsWith(prefix);
 */

File: LeetCode/0208.cpp
class Trie
{
    struct node
    {
        node *child[26];
        bool isword;
        node(bool _isword) : isword(_isword)
        {
            for (int i = 0; i < 26; i++)
                child[i] = nullptr;
        }
    };
    node *root;

public:
    /** Initialize your data structure here. */
    Trie()
    {
        root = new node(false);
    }

    /** Inserts a word into the trie. */
    void insert(string word)
    {
        auto cur = root;
        for (auto i : word)
        {
            int idx = i - 'a';
            if (cur->child[idx] == nullptr)
            {
                cur->child[idx] = new node(false);
            }
            cur = cur->child[idx];
        }
        cur->isword = true;
    }

    /** Returns if the word is in the trie. */
    bool search(string word)
    {
        auto cur = root;
        for (auto i : word)
        {
            int idx = i - 'a';
            if (cur->child[idx] == nullptr)
            {
                return false;
            }
            cur = cur->child[idx];
        }
        return cur->isword;
    }

    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix)
    {
        auto cur = root;
        for (auto i : prefix)
        {
            int idx = i - 'a';
            if (cur->child[idx] == nullptr)
            {
                return false;
            }
            cur = cur->child[idx];
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */